itest/psbt_test.go

```diff
@@ -1608,6 +1619,436 @@ func testPsbtSighashNoneInvalid(t *harnessTest) {
 	require.ErrorContains(t.t, err, "unable to verify proof")
 }
 
+// testPsbtTrustlessSwap tests that the SIGHASH_NONE flag of vPSBTs can be used
+// to execute a trustless swap between two parties. This is done by using
+// different sighashes for the bitcoin psbt and taproot asset vpsbt. One is able
+// to "claim" the assets only by bringing their own bitcoin to fulfill the
+// outputs of the bitcoin transaction.
+func testPsbtTrustlessSwap(t *harnessTest) {
+	// First, we'll make a normal asset.
+	rpcAssets := MintAssetsConfirmBatch(
+		t.t, t.lndHarness.Miner.Client, t.tapd,
+		[]*mintrpc.MintAssetRequest{issuableAssets[0]},
+	)
+
+	mintedAsset := rpcAssets[0]
+	genInfo := mintedAsset.AssetGenesis
+
+	ctxb := context.Background()
+
+	var (
+		alice       = t.tapd
+		numUnits    = uint64(5000)
+		chainParams = &address.RegressionNetTap
+		assetID     asset.ID
+	)
+	copy(assetID[:], genInfo.AssetId)
+
+	// Now Alice will derive the script and anchor internal keys that will be
+	// used to bootstrap an interactive full send of her assets.
+	aliceScriptKey, aliceAnchorInternalKey := deriveKeys(t.t, alice)
+	vPkt := tappsbt.ForInteractiveSend(
+		assetID, numUnits, aliceScriptKey, 1, aliceAnchorInternalKey,
+		asset.V0, chainParams,
+	)
+
+	// Now we fund the vPSBT, which creates 1 input and 1 output, which
+	// correspond to Alice's anchor that carries the asset and a the output to
+	// which Alice is sending all of the assets.
+	fundResp := fundPacket(t, alice, vPkt)
+
+	var err error
+	vPkt, err = tappsbt.Decode(fundResp.FundedPsbt)
+	require.NoError(t.t, err)
+
+	require.Len(t.t, vPkt.Inputs, 1)
+	require.Len(t.t, vPkt.Outputs, 1)
+
+	t.Logf("Funded vPSBT: %v", toBase64(t.t, vPkt))
+
+	// On the vPSBT level, which describes the assets transfer, we do not
+	// commit to any outputs.
+	vPkt.Inputs[0].SighashType = txscript.SigHashNone
+
+	// Let's do some sanity checks on the structure of the vPSBT, and prepare
+	// its outputs before signing.
+	require.Equal(t.t, vPkt.Outputs[0].Type, tappsbt.TypeSimple)
+	require.NoError(t.t, tapsend.PrepareOutputAssets(ctxb, vPkt))
+	require.Nil(t.t, vPkt.Outputs[0].Asset.SplitCommitmentRoot)
+	require.Len(t.t, vPkt.Outputs[0].Asset.PrevWitnesses, 1)
+	require.Nil(t.t, vPkt.Outputs[0].Asset.PrevWitnesses[0].SplitCommitment)
+
+	fundedPsbtBytes, err := tappsbt.Encode(vPkt)
+	require.NoError(t.t, err)
+
+	// Alice signs the vPSBT.
+	signedResp, err := alice.SignVirtualPsbt(
+		ctxb, &wrpc.SignVirtualPsbtRequest{
+			FundedPsbt: fundedPsbtBytes,
+		},
+	)
+	require.NoError(t.t, err)
+	require.Contains(t.t, signedResp.SignedInputs, uint32(0))
```

Comment:
`txscript.SigHashNone` is used so that we can sign the input and allow its spending while also committing to none of the outputs, allowing somebody (Bob) later to edit the outputs.

If we isolate on the vPSBT level, Alice is basically saying "i'm signing these assets, and anyone can define the output"

Reply:
Thanks for the explanation!